ml-proto/spec/int.ml

```diff
@@ -208,7 +210,49 @@ struct
   let ge_s x y = x >= y
   let ge_u x y = cmp_u x (>=) y
 
-  let of_string = Rep.of_string
+  let parse_hexdigit c =
+    int_of_char c -
+      if '0' <= c && '9' >= c then 0x30
+      else if 'a' <= c && 'f' >= c then 0x61 - 10
+      else if 'A' <= c && 'F' >= c then 0x41 - 10
+      else failwith "unexpected digit"
+
+  let neg_safe x =
+    if lt_s (sub x Rep.one) Rep.minus_one then raise Numerics.IntegerOverflow;
+    Rep.neg x
+
+  (* This implementation allows leading signs and unsigned values *)
+  let of_string x =
+    let open Rep in
+    let len = String.length x in
+    let require b = if not b then failwith "of_int" in
+    let rec parse_hex i num =
+      if i = len then num
+      else begin
+        require (le_u num (shr_u minus_one (of_int 4)));
+        parse_hex (i + 1) (logor (shift_left num 4) (of_int (parse_hexdigit x.[i])))
+      end
+    in
+    let rec parse_dec i num =
+      if i = len then num
+      else begin
+        require ('0' <= x.[i] && '9' >= x.[i]);
+        let new_digit = of_int (int_of_char x.[i] - 0x30) in
```

Comment:
Just for symmetry with `parse_hexdigit`, perhaps move this and the previous line to a helper function, too.